clash-prelude/src/Clash/Clocks/Deriving.hs

```diff
@@ -78,6 +78,16 @@ derive' n = do
       let c = varT $ mkName ("c" ++ show n') in
       [t| Clock $c |]
 
+    -- | Generate @!_ = knownDomain dom@ with fresh @dom@ variable
```

Comment:
Ahh, now I get it. It's the `unsafeCoerce` that is wreaking havoc; and `Clash.Intel.ClockGen.altpll` prevents the `KnownDomain`s from being eliminated itself, which is why that one does work. But the only reason for the `unsafeCoerce` was to avoid `BangPatterns` which is enabled by default anyways. Plus it is no longer needed because we pattern match to detect dynamic clocks. And to top it off you're introducing bang patterns yourself here. So let's just get rid of the `unsafeCoerce` instead:

```diff
diff --git a/clash-prelude/src/Clash/Clocks/Deriving.hs b/clash-prelude/src/Clash/Clocks/Deriving.hs
index 8553e8f69..dfe62c369 100644
--- a/clash-prelude/src/Clash/Clocks/Deriving.hs
+++ b/clash-prelude/src/Clash/Clocks/Deriving.hs
@@ -13,12 +13,12 @@ Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
 module Clash.Clocks.Deriving (deriveClocksInstances) where
 
 import Control.Monad               (foldM)
+import Clash.Promoted.Symbol       (SSymbol(..))
 import Clash.Explicit.Signal       (unsafeSynchronizer)
 import Clash.Signal.Internal
 import Language.Haskell.TH.Compat
 import Language.Haskell.TH.Syntax
 import Language.Haskell.TH.Lib
-import Unsafe.Coerce               (unsafeCoerce)
 
 conPatternNoTypes :: Name -> [Pat] -> Pat
 #if MIN_VERSION_template_haskell(2,18,0)
@@ -46,24 +46,22 @@ derive' n = do
 #endif
 
   -- Function definition of 'clocks'
-  let clk = mkName "clk"
   let rst = mkName "rst"
 
   -- Implementation of 'clocks'
+  clkImpl <- [| Clock SSymbol Nothing |]
   lockImpl <- [| unsafeSynchronizer clockGen clockGen
                    (unsafeToLowPolarity $(varE rst)) |]
   let
     noInline  = PragmaD $ InlineP (mkName "clocks") NoInline FunLike AllPhases
-    clkImpls  = replicate n (clkImpl clk)
+    clkImpls  = replicate n clkImpl
     instTuple = mkTupE $ clkImpls ++ [lockImpl]
     funcBody  = NormalB instTuple
     errMsg    = "clocks: dynamic clocks unsupported"
     errBody   = NormalB ((VarE 'error) `AppE` (LitE (StringL errMsg)))
     instFunc  = FunD (mkName "clocks")
       [ Clause
-          [ AsP
-              clk
-              (conPatternNoTypes 'Clock [WildP, conPatternNoTypes 'Nothing []])
+          [ (conPatternNoTypes 'Clock [WildP, conPatternNoTypes 'Nothing []])
           , VarP rst]
           funcBody
           []
@@ -91,9 +89,6 @@ derive' n = do
       let p = varT $ mkName "pllLock" in
       [t| KnownDomain $p |]
 
-
-    clkImpl clk = AppE (VarE 'unsafeCoerce) (VarE clk)
-
 -- Derive instances for up to and including to /n/ clocks
 deriveClocksInstances :: Int -> Q [Dec]
 deriveClocksInstances n = mapM derive' [1..n]
```
(this diff is against the original code of this file for clarity).

Your solution boils down to:

```haskell
instance Clocks (Clock c1, Clock c2, Signal pllLock Bool) where

  type ClocksCxt (Clock c1, Clock c2, Signal pllLock Bool) =
    (KnownDomain c1, KnownDomain c2, KnownDomain pllLock)

  clocks clk@(Clock _ Nothing) rst =
    ( unsafeCoerce clk
    , unsafeCoerce clk
    , unsafeSynchronizer clockGen clockGen (unsafeToLowPolarity rst))
   where
    !_ = knownDomain @c1
    !_ = knownDomain @c2
  clocks _ _ = error "clocks: dynamic clocks unsupported"
  {-# NOINLINE clocks #-}
```

while mine is
```haskell
instance Clocks (Clock c1, Clock c2, Signal pllLock Bool) where

  type ClocksCxt (Clock c1, Clock c2, Signal pllLock Bool) =
    (KnownDomain c1, KnownDomain c2, KnownDomain pllLock)

  clocks (Clock _ Nothing) rst =
    ( Clock SSymbol Nothing
    , Clock SSymbol Nothing
    , unsafeSynchronizer clockGen clockGen (unsafeToLowPolarity rst))
  clocks _ _ = error "clocks: dynamic clocks unsupported"
  {-# NOINLINE clocks #-}
```